code/drasil-database/lib/Database/Drasil/ChunkDB.hs

```diff
@@ -229,6 +241,82 @@ insertAll as cdb = foldl' (flip insert) cdb as
 -- built properly (i.e., using the `HasChunkRefs` typeclass).
 --------------------------------------------------------------------------------
 
+-- | Insert 12 lists of /unique/ chunk types into a 'ChunkDB', assuming the
+-- input 'ChunkDB' does not already contain any of the chunks from the chunk
+-- lists.
+--
+-- Warning: Ignores the dependencies of 'LabelledContent's and 'Reference's
+-- (because they [a] should have none and [b] don't matter enough to care for).
+insertAllOutOfOrder12 ::
+  (IsChunk a, IsChunk b, IsChunk c, IsChunk d, IsChunk e,
+   IsChunk f, IsChunk g, IsChunk h, IsChunk i, IsChunk j) =>
+   ChunkDB ->
+   [a] -> [b] -> [c] -> [d] -> [e] ->
+   [f] -> [g] -> [h] -> [i] -> [j] ->
+   [LabelledContent] -> [Reference] -> ChunkDB
+insertAllOutOfOrder12 strtr as bs cs ds es fs gs hs is js lcs rs =
+  let
+    -- Box all of our chunks
+    as' = map mkChunk as
+    bs' = map mkChunk bs
+    cs' = map mkChunk cs
+    ds' = map mkChunk ds
+    es' = map mkChunk es
+    fs' = map mkChunk fs
+    gs' = map mkChunk gs
+    hs' = map mkChunk hs
+    is' = map mkChunk is
+    js' = map mkChunk js
+
+    -- Put all of our chunks in a list of lists, with each list carrying a
+    -- unique type of chunk, filtering out empty lists
+    altogether = filter (not . null) 
+                  [as', bs', cs', ds', es',
+                   fs', gs', hs', is', js']
+    calt = concat altogether
+
+    -- Calculate the chunk dependancies (i.e., UID -> Dependants)
+    chDeps = invert $ M.fromList $ map (\c -> (c ^. uid, S.toList $ chunkRefs c)) calt
+    hardLookup k m = fromMaybe (error "this situation will never happen") $ M.lookup k m
+
+    -- Create the chunk table for the incoming chunks
+    chDepsTab = M.fromList ((\c -> (c ^. uid, (c, hardLookup (c ^. uid) chDeps))) <$> calt)
+
+    -- Insert all incoming chunks with the existing chunk table, asserting that
+    -- none of the inserted chunks were already inserted.
+    chTab = M.unionWith (\(x, _) _ -> error $ "duplicate chunk found in mass insertion: " ++ show (x ^. uid))
+      (chunkTable strtr) (M.fromList $ map (\c -> (c ^. uid, (c, []))) calt)
```

Comment:
To be perfectly clear, it _is necessary_ to add the new chunks in two phases: adding with an empty list of dependancies and then adding the list of dependencies because we need to make sure we're not overwriting any chunks of different types. Perhaps the code is just 'easier to write' this way and is not the 'most efficient' this way, however...

I think we would need 'deep equality' for the more memory-efficient version (i.e., the version that does not create two maps with more or less the same contents excluding the dependancies.